Random

TyPR124 opened the pull request:
Let me know what you think of this. It should handle the case of rdrand instruction failing. It also makes the RdRand::get() method generic, so it can return a u64, u32, or u16, depending on what the caller wants.

I removed the target requirement on the 'get' method, since x86 can use the instruction for u16 and u32, just not u64. I put the x86_64 target requirement on the u64 RdRandPrimitive.

I did NOT remove the x86_64 target requirement from RdRand::new(), since it appears the raw_cpuid dependency is only used in x86_64. If the dependency can be added for other targets, then we can remove the requirement from RdRand::new() as well.

I've tested this with a release build, and confirmed that the assembly output picks the correct register size in each scenario.

I have not tested a failure scenario. I may try to do this later. I'm thinking if I run the instruction a whole bunch of times in parallel, hopefully I can get a failure to happen.

I may also play with the idea of something like `pub fn fill(&self, &mut T) -> bool` which would let rdrand write directly into the destination.

phil-opp commented on src/random/mod.rs:
```diff
@@ -20,15 +30,35 @@ impl RdRand {
+    }
+    /// Uniformly sampled u32
+    #[inline]
+    pub fn get_u32(&self) -> Option<u32> {
```
This function should have `#[cfg(target_arch = "x86_64")]` too. It might work on x86, but fails on non-x86 platforms (this crate should be buildable on all platforms).

While `#[cfg(target_arch = "x86")]` would work too, I don't think that it's a good idea to add support for 32-bit targets because the Rest of the crate knows only two modes: x86_64 and non-x86_64.

TyPR124 replied:
This makes sense. I've pulled x86_64 requirements up to the impl level, since everything in the impl would need it anyway.

phil-opp replied:
Thanks!

phil-opp commented on src/random/mod.rs:
```diff
@@ -20,15 +30,35 @@ impl RdRand {
+    }
+    /// Uniformly sampled u16
+    #[inline]
+    pub fn get_u16(&self) -> Option<u16> {
```
Same as above.

TyPR124 commented on src/random/mod.rs:
```diff
@@ -1,12 +1,22 @@
+    #[cfg(target_arch = "x86_64")]
+    impl RdRandPrimitive for u64 {}
+}
+use private::RdRandPrimitive;
```
So RdRandPrimitive is okay this way because it is still technically a public thing, just inside a non-public module. I've seen this done in other crates, and also recommended on the rust subreddit as a way to prevent implementing this trait outside of this crate.

Essentially, if RdRandPrimitive is accessible to users of this crate, they could write something like `struct Thing(u64); impl RdRandPrimitive for Thing{}`. This specific example is okay, but if they changed it to `struct Thing(u128); impl RdRandPrimitive for Thing{}`, then that is NOT okay, but still compiles without warning. The latter is not okay because rdrand does not output a 128bit value, so part of `Thing` will be would end up being non-random.

The solutions to this (maybe there's more, just what I could think of) are to either make the trait unsafe (and rely on the implementor to understand how to use it correctly), or put the trait in a private module so the user cannot implement it, and can't do something dumb like the example above.

phil-opp replied:
I didn't know that this was a common thing. It still seems hacky and leads to an ugly API documentation:

![image](https://user-images.githubusercontent.com/1131315/52519623-6d12b000-2c5e-11e9-987c-f3cec323561d.png)

There is no way for the user to see for which types `RdRandPrimitive` is implemented because there's no documentation for it as it lives in a private module.

TyPR124 replied:
I can make documentation more explicit, or pull the trait out of the private module and use a private "Sealed" pattern like they do in in link.

TyPR124 commented:
I've made some changes. Restored the original `get` and deprecated it. I renamed the generic get to `rand`. And I fixed the x86_64 requirements like you've suggested.

I do want to make an argument in favor of leaving the generic `rand` fn. If you still think it is too much, I can remove it.

By doing this, we only have to maintain one block of assembly (and the surrounding logic), as opposed to three blocks that are all the same. Similarly, if we implement RdSeed using a similar pattern, then we'll have 2 blocks of assembly to maintain instead of 6.

And even though the trait is (effectively) private, the caller can still just call `rng.rand::<u64>()` (for example), so that part doesn't add any extra complexity.

About possibly having to specify type parameters, I personally don't mind that. I think `rng.rand::<u32>()` is perfectly readable, but that is why I also included the `get_u*` fns as well, because I know it isn't for everybody. Another benefit is that if users are relying heavily on type inference, they can likely get away with just `rng.rand()` and then freely change between u16, u32, or u64 if they need to, without having to change that specific function call.

TyPR124 commented:
Yea, I also looked for register constraints of specific sizes, and it doesn't seem like that exists.

If it helps at all, this is how the official Intel library does it.

`#  define _rdrand_step(x) ({ unsigned char err; asm volatile("rdrand %0; setc %1":"=r"(*x), "=qm"(err)); err; })

#  define _rdrand16_step(x) _rdrand_step(x)
#  define _rdrand32_step(x) _rdrand_step(x)`

https://software.intel.com/en-us/articles/the-drng-library-and-manual#

My understanding is that rusts asm! should behave exactly like the C asm, so this shouldn't break.

TyPR124 commented:
How would you feel about having some included tests that check if the destination was properly filled with random data? The main downside is I can't immediately think of a way to do this without using 0 (a possibly valid random value) for failure checks, which could lead (randomly, probably 1 in 65,000 tests) to false failures.

phil-opp commented:
Maybe it won't break, but it's difficult to be sure without an exact specification of inline assembly.

I think just copying the `asm` statement three times is both simpler and safer:

- It's apparent what the functions do without understanding the (not really documented) details of inline assembly.
- We can be reasonably sure that it won't break in the future without needing tests.
- There's no chance that someone introduces a `get_u128` method in the future without noticing that it doesn't work (because there are no warnings).
- We don't need the hidden `RdRandPrimitive` trait.
- It's just two additional `asm` instructions that we probably won't touch again. And even if we need to, we only need to edit three lines instead of one, which is manageable.

So I think the advantages outweigh the bit of code duplication.

TyPR124 commented:
Fair enough, I'll change it. It will the downside of not letting the compiler pick a register for the random data. I can still use the `r` constraint for `setc`, since the instruction only  works with a byte destination, unless you would rather that be explicit also.

phil-opp commented:
Thanks!

> It will the downside of not letting the compiler pick a register for the random data.

You're right, that's a downside. But don't think that it leads to performance problems since `rdrand` is a slow operation.

> I can still use the `r` constraint for `setc`, since the instruction only works with a byte destination, unless you would rather that be explicit also.

A `r` constraint for `setc` is fine.

TyPR124 commented:
It is updated. Let me know if you see any problems.

phil-opp commented:
Thanks a lot!

phil-opp commented:
Published as version 0.4.2
